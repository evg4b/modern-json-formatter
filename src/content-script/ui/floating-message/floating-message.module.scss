:host {
  display: flex;
  position: fixed;
  right: 25px;
  bottom: 15px;
  padding: 10px;
  background: #282828;
  color: #eee;
  box-shadow: 0 5px 10px 0 rgb(0 0 0 / 70%);
  border-radius: 5px;
  font-size: 10px;
  flex-direction: column;
  max-width: 300px;
  user-select: none;
  opacity: 1;
  transform: translateY(0);
  transition: all 250ms ease-in-out;

  .header-container {
    display: flex;
    flex-direction: row;

    .header {
      flex: 1 1 auto;
      margin-bottom: 5px;
      font-size: 12px;
    }

    .close {
      cursor: pointer;
      width: 12px;
      height: 12px;
      position: relative;

      &:before,
      &:after {
        position: absolute;
        width: 2px;
        background-color: #eee;
        left: 5px;
        content: " ";
        height: 12px;
      }

      &:before {
        transform: rotate(45deg);
      }

      &:after {
        transform: rotate(-45deg);
      }
    }
  }

  .body {
    display: flex;
    flex-direction: column;
    color: #9b9b9b;
  }
}

:host.hidden,
:host-context(.hidden) {
  opacity: 0;
  transform: translateY(50%);
}
