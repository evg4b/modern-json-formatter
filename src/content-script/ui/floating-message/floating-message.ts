import { createElement, CustomElement, StyledComponentElement } from '@core/dom';
import floatingMessageStyles from './floating-message.styles.txt';

@CustomElement('floating-message')
export class FloatingMessageElement extends StyledComponentElement {
  private readonly close = createElement({ element: 'div', class: 'close' });

  constructor(header: string, message: string) {
    super(floatingMessageStyles);
    this.classList.add('hidden');

    this.shadow.append(
      createElement({
        element: 'div',
        class: 'header-container',
        children: [
          createElement({ element: 'div', class: 'header', content: header }),
          this.close,
        ],
      }),
      createElement({ element: 'div', class: 'body', content: message }),
    );

    setTimeout(() => this.classList.remove('hidden'), 10);

    const id = setTimeout(() => this.closeMessage(), 10_000);
    this.close.addEventListener('click', () => {
      clearTimeout(id);
      this.closeMessage();
    });
  }

  private closeMessage() {
    this.classList.add('hidden');
    setTimeout(() => {
      this.remove();
    }, 250);
  }
}
