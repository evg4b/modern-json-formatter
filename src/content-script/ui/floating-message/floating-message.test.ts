jest.useFakeTimers();

import { afterEach, beforeEach, describe, expect, jest, test } from '@jest/globals';
import { FloatingMessageElement } from './floating-message';

describe('FloatingMessageElement', () => {
  let floatingMessage: FloatingMessageElement;

  beforeEach(() => {
    floatingMessage = new FloatingMessageElement('Test Header', 'Test Message');
    document.body.appendChild(floatingMessage);
  });

  afterEach(() => {
    document.body.removeChild(floatingMessage);
  });

  test('should create a hidden floating message element', () => {
    expect(floatingMessage.classList).toContain('hidden');
  });

  test('de', () => {
    expect(document.body.firstChild).toBe(floatingMessage);
  });

  describe('header', () => {
    beforeEach(() => {
      jest.advanceTimersByTime(1000);
    });

    test('should display the message after a timeout', () => {
      expect(floatingMessage.classList).not.toContain('hidden');
    });

    test('should display the header', () => {
      expect(floatingMessage.shadowRoot).toMatchSnapshot();
    });
  });
});
