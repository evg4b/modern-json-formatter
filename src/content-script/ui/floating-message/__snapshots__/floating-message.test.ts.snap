// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FloatingMessageElement header should display the header 1`] = `null`;
